.config/doom/config.org
#+title: Doom literate config

* File Headline
#+begin_src emacs-lisp :tangle yes
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src
* User informations
#+begin_src emacs-lisp :tangle yes
(setq user-full-name "Frédéric Willem"
      user-mail-address "frederic.willem@gmail.com")
#+end_src
* Calendar/holidays
#+begin_src emacs-lisp :tangle yes
(setq calendar-location-name "Saint-Nicolas, BE"
      calendar-latitude 50.628
      calendar-longitude 5.516)
(custom-set-variables
 '(holiday-bahai-holidays nil)
 '(holiday-hebrew-holidays nil)
 '(holiday-islamic-holidays nil))
#+end_src


* UI and appearance
#+begin_src emacs-lisp :tangle yes

(setq doom-font "FiraCode Nerd Font-12")
;; doom-symbol-font "Nerd Font Symbol")

(setq doom-theme 'modus-vivendi-tinted)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

(+global-word-wrap-mode +1)
#+end_src
* Org mode
If you use `org' and don't want your org files in the default location below, change `org-directory'. It must be set before org loads!
#+begin_src emacs-lisp :tangle yes
(setq org-directory "~/org/")
(setq org-agenda-files (list "~/org/gtd"))
(setq denote-directory org-directory)
(setq org-passwords-file "~/org/password.org.gpg")
(map!
 :leader
 :desc "Org passwords" "o p" #'org-passwords)
(setq org-hide-emphasis-markers t)
(after! org
  ;; for org capture extension
  (require 'org-protocol)
  ;; needed by org-contacts
  (require 'ol)
  (setq org-todo-keywords
        '(
          (sequence "TODO(t)" "NEXT(n)" "WAIT(w)" "|" "DONE(d)" "CNCLD(c@)")
          (sequence "PROJECT(p)" "|" "KILLED(k@)")))
  (setq org-tag-alist '(("@work") ("@home") ("@phone") ("@computer") ("@online") ("@errand")))
  (setq org-gtd-inbox-file "~/org/gtd/0-inbox.org")
  (add-to-list 'org-capture-templates
               '("i" "Inbox" entry
                 (file org-gtd-inbox-file)
                 "* %?\n%i\n%a"))
  (add-to-list 'org-capture-templates
               '("p" "Protocol" entry (file org-gtd-inbox-file)
                 "* %^{Title}\nSource: %u, %c\n #+BEGIN_QUOTE\n%i\n#+END_QUOTE\n\n\n%?") t)
  (add-to-list 'org-capture-templates
               '("L" "Protocol Link" entry (file org-gtd-inbox-file)
                 "* %? [[%:link][%:description]] \nCaptured On: %U") t)

  (delete '("p" "Templates for projects")
          org-capture-templates)
  (delete '("pn" "Project-local notes" entry
            (file+headline +org-capture-project-notes-file "Inbox")
            "* %U %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete '("n" "Personal notes" entry
            (file+headline +org-capture-notes-file "Inbox")
            "* %u %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete '("pt" "Project-local todo" entry
            (file+headline +org-capture-project-todo-file "Inbox")
            "* TODO %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete '("oc" "Project changelog" entry #'+org-capture-central-project-changelog-file "* %U %?\n %i\n %a" :heading "Changelog" :prepend t)
          org-capture-templates)
  (delete '("on" "Project notes" entry #'+org-capture-central-project-notes-file "* %U %?\n %i\n %a" :heading "Notes" :prepend t)
          org-capture-templates)
  (delete '("o" "Centralized templates for projects")
          org-capture-templates)
  (delete '("t" "Personal todo" entry
            (file+headline +org-capture-todo-file "Inbox")
            "* [ ] %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete  '("pc" "Project-local changelog" entry
             (file+headline +org-capture-project-changelog-file "Unreleased")
             "* %U %?\n%i\n%a" :prepend t)
           org-capture-templates)
  (delete '("ot" "Project todo" entry #'+org-capture-central-project-todo-file "* TODO %?\n %i\n %a" :heading "Tasks" :prepend nil)
          org-capture-templates))
(setq org-agenda-custom-commands
      '(("n" "Agenda and all NEXT tasks"
         ((agenda "" nil)
          (todo "NEXT" nil))
         nil)))
(defvar org-gtd-archive-file "~/org/gtd/_gtd_archive_2024")
(setq org-archive-location (concat org-gtd-archive-file "::datetree/"))
(setq org-log-done 'time)
#+end_src

* General defaults
#+begin_src emacs-lisp :tangle yes
(add-hook 'after-save-hook
          'executable-make-buffer-file-executable-if-script-p)
(setq bookmark-default-file "~/.config/doom/bookmarks" )
#+end_src

* Completions
#+begin_src emacs-lisp :tangle yes
(setq projectile-project-search-path '(( "~/Repos/" . 1 )))
(after! orderless
  (add-to-list 'completion-styles 'flex t))
(after! corfu
  (setq corfu-preselect 'directory))
#+end_src
* Mail config
- [[https://github.com/gauteh/lieer][lieer]] is used to communicate with gmail
  #+begin_src shell :tangle no
paru -S lieer-git
  #+end_src
- [[https://gitlab.com/goobook/goobook][GooBook]] is used for contact management
  #+begin_src shell :tangle no
paru -S goobook
  #+end_src
#+begin_src emacs-lisp :tangle yes
(require 'notmuch-mua)
(global-set-key [remap compose-mail] #'+notmuch/compose)
(after! notmuch
  (set-popup-rule! "^\\*notmuch-hello" :ignore t)
  (setq notmuch-multipart/alternative-discouraged '("text/plain" "multipart/related"))
  (setq +notmuch-mail-folder "~/.mail/")
  (setq sendmail-program "gmi")
  (setq message-sendmail-extra-arguments '("send" "--quiet" "-t" "-C" "~/.mail"))
  (setq send-mail-function 'sendmail-send-it)
  (setq message-send-mail-function 'message-send-mail-with-sendmail)
  (require 'notmuch-address)
  (setq notmuch-address-command "~/.scripts/gook")
  (map!
   :map (notmuch-search-mode-map notmuch-tree-mode-map notmuch-show-mode-map)
   :nv [remap evil-undo] #'notmuch-tag-undo)

  (defun +notmuch/search-trash ()
    "Flag the message."
    (interactive)
    (notmuch-search-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "trash" "search" #'notmuch-tree-next-message))
  (defun +notmuch/show-trash ()
    "Flag the message."
    (interactive)
    (notmuch-show-remove-tag'("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "trash" "show" #'notmuch-show-next-thread-show))
  (defun +notmuch/tree-trash ()
    "Flag the message."
    (interactive)
    (notmuch-tree-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "trash" "tree"))
  (global-set-key [remap evil-collection-notmuch-tree-toggle-delete] #'+notmuch/tree-trash)
  (global-set-key [remap evil-collection-notmuch-show-toggle-delete] #'+notmuch/show-trash)
  (global-set-key [remap evil-collection-notmuch-search-toggle-delete] #'+notmuch/search-trash)
  (defun +notmuch/search-flagged ()
    "Flag the message."
    (interactive)
    (notmuch-search-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "flagged" "search" #'notmuch-tree-next-message))
  (defun +notmuch/show-flagged ()
    "Flag the message."
    (interactive)
    (notmuch-show-remove-tag'("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "flagged" "show" #'notmuch-show-next-thread-show))
  (defun +notmuch/tree-flagged ()
    "Flag the message."
    (interactive)
    (notmuch-tree-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "flagged" "tree"))
  (global-set-key [remap evil-collection-notmuch-tree-toggle-flagged] #'+notmuch/tree-flagged)
  (global-set-key [remap evil-collection-notmuch-show-toggle-flagged] #'+notmuch/show-flagged)
  (global-set-key [remap evil-collection-notmuch-search-toggle-flagged] #'+notmuch/search-flagged))
#+end_src
* Spelling
Spelling: avoid tis

https://github.com/minad/jinx/
Faster and support multiple dictionaries out-of-the-box
#+begin_src emacs-lisp :tangle yes
(use-package! jinx
  :general ([remap ispell-word] #'jinx-correct)
  :init
  (after! evil
    (global-set-key [remap evil-next-flyspell-error] #'jinx-next)
    (global-set-key [remap evil-prev-flyspell-error] #'jinx-previous))
  (setq jinx-languages "fr_FR en_US"))

(add-hook! (prog-mode text-mode) #'jinx-mode )
#+end_src

* Evil
#+begin_src emacs-lisp :tangle yes
(after! evil
  (defun +evil-paste-above ()
    "Paste the line above."
    (interactive)
    (evil-insert-newline-above)
    (evil-paste-after 1 evil-this-register))

  (defun +evil-paste-below ()
    "Paste the line below."
    (interactive)
    (evil-insert-newline-below)
    (evil-paste-after 1 evil-this-register))
  (map!
   :map evil-normal-state-map
   "[ p" #'+evil-paste-above
   "] p" #'+evil-paste-below ))
#+end_src

* Org contacts
#+begin_src emacs-lisp :tangle yes
(setq org-contacts-files `(,(expand-file-name (concat org-directory "contacts.org"))
                           ,(expand-file-name (concat org-directory "contacts-maman.org"))))
(setq org-contacts-matcher
      "N<>\"\"|EMAIL<>\"\"|ALIAS<>\"\"|PHONE<>\"\"|ADDRESS<>\"\"|BIRTHDAY<>\"\"")
(add-load-path! "~/.config/doom/lisp/")
(require 'gtd)
(use-package! org-contacts
  :commands org-contacts-agenda)
;; (map!
;;  :mode ( text-mode prog-mode )
;;  :gi "TAB"   #'yasnippet-capf
;;  :gi "<tab>" #'yasnippet-capf)
(map!
 :leader
 :desc "Org contacts" "o c" #'org-contacts-agenda)
#+end_src

#+begin_src emacs-lisp :tangle yes

;; Bind your key
(map! :map notmuch-search-mode-map
      :n "K" #'notmuch-tag-jump)
(map! :map dired-mode-map
      :n "K" #'dired-do-kill-lines)
;; Optionally re-bind documentation to different key:
(map! :nv "gK"  #'+lookup/documentation)
(defhydra hydra-smartparens ()
  "Smartparens"
  ("q" nil)

  ;; Wrapping
  ("(" (lambda (_) (interactive "P") (sp-wrap-with-pair "(")))
  ("{" (lambda (_) (interactive "P") (sp-wrap-with-pair "{")))
  ("'" (lambda (_) (interactive "P") (sp-wrap-with-pair "'")))
  ("\"" (lambda (_) (interactive "P") (sp-wrap-with-pair "\"")))

  ("w" (lambda (_) (interactive "P") (sp-wrap-with-pair "(")) "wrap")
  ("W" sp-unwrap-sexp)

  ;; Movement
  ("l" sp-next-sexp)
  ("h" sp-backward-sexp)
  ("j" sp-down-sexp)
  ("k" sp-backward-up-sexp)

  ("L" sp-forward-symbol)
  ("H" sp-backward-symbol)

  ("^" sp-beginning-of-sexp)
  ("$" sp-end-of-sexp)

  ("t" sp-transpose-sexp "transpose")
  ("u" undo-tree-undo "undo")

  ("y" sp-copy-sexp "copy")
  ("d" sp-kill-sexp "delete")

  ("s" sp-forward-slurp-sexp "slurp")
  ("S" sp-backward-slurp-sexp)

  ("b" sp-forward-barf-sexp "barf")
  ("B" sp-backward-barf-sexp)

  ("v" sp-select-next-thing "select")
  ("V" sp-select-previous-thing))
(map! :localleader
        :mode writegood-mode
        "g" #'writegood-grade-level
        "r" #'writegood-reading-ease)
#+end_src


history.txt
sudo apt show lazygal
lazygit
cd .dotfiles/
neomutt
mbsync -a
notmuch new
notmuch
sudo apt install notmuch
exit
nvim .pam-gnupg
gpg -K --with-keygrip
nvim .gnupg/gpg-agent.conf
cd
sudo nvim /etc/pam.d/login
sudo make install
make
./configure
sudo apt install libpam0g-dev
./autogen.sh
sudo apt-get install ninja-build gettext libtool libtool-bin autoconf automake cmake g++ pkg-config unzip curl doxygen build-essential
./autogen.sh
./configure
make
make install
cd pam-gnupg/
git clone https://github.com/cruegge/pam-gnupg.git
meteo_full
meteo
test
idea.sh &> /dev/null &
fish_add_path /home/willefi/.local/opt/idea-IC-223.7571.182/bin/
tar xzvf ideaIC-2022.3.tar.gz
cd .local/opt/
mv -iv ideaIC-2022.3.tar.gz .local/opt/
md .local/opt
rm -irv ideaIC-2022.3.exe
mv -iv /mnt/c/Users/User/Downloads/ideaIC-2022.3.tar.gz .
mv -iv /mnt/c/Users/User/Downloads/ideaIC-2022.3.exe .
wget https://download.jetbrains.com/idea/ideaIC-2022.3.tar.gz?_gl=1*1o7ww63*_ga*ODgyNzc5MTU5LjE2NzA2Njg1MTM.*_ga_9J976DJZ68*MTY3MDg3MDU0MS44LjEuMTY3MDg3MDU3OS4wLjAuMA..&_ga=2.174583842.1298344305.1670668513-882779159.1670668513
pass show frederic.willem@gmail.com
pass shox frederic.willem@gmail.com
pass insert frederic.willem@gmail.com
nvim willydeliege/Configure\ mutt\ for\ gmail.md
pass insert
pass init frederic.willem@gmail.com
sudo apt install pass
md /home/willefi/.local/share/mail/frederic.willem@gmail.com
sudo apt install isync
gpg --import p.k
git checkout develop
git restore .
cd willydeliege/
ls
sudo apt install neomutt  msmtp
sudo apt install neomutt mbsync msmtp
stow neomutt/
xdg-open http://google.be
pip install --user git+https://github.com/cpbotha/xdg-open-wsl.git
df -h --total
df -h totol
du
rm -rf gitsign_0.3.0_linux_amd64.deb
ll
install --help  &| less
install --help
install
htop
dictionary --version
dictionary -v
dictionary -h
dictionary --help
dictionary word
cargo install dictionary
sudo apt install libssl-dev
nvimrc
fishrc
gitsign --help
gitsign
sudo dpkg -S dictionary
sudo dpkg -S dictionnary
wget https://github.com/sigstore/gitsign/releases/download/v0.3.0/gitsign_0.3.0_linux_amd64.deb
sudo dpkg -i gitsign_0.3.0_linux_amd64.deb
cargo install shellharden
sudo apt install zsh
shellharden
sudo apt install black isort
isort
black
flake8
sudo apt install python3-proselint
sudo apt install pythoon3-proselint
proselint
sudo apt install node-typescript
tsc
npm install tsc
pip install flake8
nvim .config/nvim/lua/plugins.lua
sudo apt remove xclip
sudo apt install xclip
sudo nvim willefi
whoami
sudo mv -iv `whoami` willefi
mv -iv `whoami` xillefi
cd /etc/sudoers.d/
echo "`whoami` ALL=(ALL) NOPASSWD:ALL" | sudo tee /etc/sudoers.d/`whoami` && sudo chmod 0440 /etc/sudoers.d/`whoami`
sudo nvim /etc/sudoers
apt install cpanminus
history
git status
j wil
sudo apt install golang
sudo gem install neovim
gem install neovim
sudo cpanm -n Neovim::Ext
sudo apt install fzf
cpanm -n Neovim::Ext
sudo apt install cpanminus
sudo apt install ruby-dev
sudo npm install -g neovim
npm install -g neovim
nvim
zk index
zk
sudo cp -irv zk /usr/local/bin/
cp -irv zk /usr/local/bin/
tar xzvf zk-v0.12.0-linux-amd64.tar.gz
sudo cp -irv bin/exa /usr/local/bin/
cp -irv bin/exa /usr/local/bin/
man exa
unzip exa-linux-x86_64-v0.10.1.zip
sudo apt install unzip
apt install unzip
tar xzvf exa-linux-x86_64-v0.10.1.zip
cd .dotfiles/librairies/
cd node_modules/
sdk install java
sdk
nvim +PackerSync
sudo apt install taskwarrior
apt install taskwarrior
pip install tasklib
pip install pynvim
npm --version
sudo npm install -g npm@9.2.0
npm install -g npm@9.2.0
npm install §g npm@9.2.0
npm install neovim
apt show npm
npm -version
sudo apt install nodejs
sudo apt remove npm nodejs
sudo apt install npm
apt install npm
sudo apt autoremove
sudo apt remove nodejs
curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash - &&\
sudo apt-get install -y nodejs
nodejs --version
nodejs -version
nodejs =version
env
sudo apt install cargo ruby
sudo apt install python3-pip
git clone https://github.com/tmux-plugins/tpm ~/.tmux/plugins/tpm
sudo apt install nodejs npm
sudo apt install neovim
tmuxrc
sudo apt install zoxide fortunes cowsay
fisher list
fisher install
stow tmux
stow fish
stow nvim
rm -rf .tmux
rm -rf .config/fish/
sudo apt install stow
sudo add-apt-repository ppa:neovim-ppa/unstable
sudo apt update
curl -sL https://git.io/fisher | source && fisher install jorgebucaran/fisher
chsh -s /usr/bin/fish
gh repo clone willydeliege/dotfiles .dotfiles
sudo apt install tmux
sudo apt install fish
sudo apt install fish xdg-utils
sudo apt-add-repository ppa:fish-shell/release-3
sudo apt install fish  tmux
sudo apt install fish fisher tmux
gh repo clone willydeliege/willydeliege
gh auth login
sudo apt install gh
xdg-open
sudo apt install wslu
sudo apt update && sudo apt-upgrade -y


doom/.config/doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq user-full-name "Frédéric Willem"
      user-mail-address "frederic.willem@gmail.com")

(setq calendar-location-name "Saint-Nicolas, BE"
      calendar-latitude 50.628
      calendar-longitude 5.516)
(custom-set-variables
 '(holiday-bahai-holidays nil)
 '(holiday-hebrew-holidays nil)
 '(holiday-islamic-holidays nil))

(setq doom-font "FiraCode Nerd Font-12")
;; doom-symbol-font "Nerd Font Symbol")

(setq doom-theme 'modus-vivendi-tinted)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

(+global-word-wrap-mode +1)
;; Colemak friendly au-keys
(after! ace-window
  (setq aw-keys '(?a ?r ?s ?t ?n ?e ?i ?o)))

(setq org-directory "~/org/")
(setq org-agenda-files (list "~/org/gtd"))
(setq org-passwords-file "~/org/password.org.gpg")
(map!
 :leader
 :desc "Org passwords" "o p" #'org-passwords)
(setq org-hide-emphasis-markers t)
(after! org
  ;; for org capture extension
  (require 'org-protocol)
  ;; needed by org-contacts
  (require 'ol)
  (setq org-todo-keywords
        '(
          (sequence "TODO(t)" "NEXT(n)" "WAIT(w)" "|" "DONE(d)" "CNCLD(c@)")
          (sequence "PROJECT(p)" "|" "KILLED(k@)")))
  (add-to-list 'org-todo-keyword-faces '("NEXT" . +org-todo-active))
  (add-to-list 'org-todo-keyword-faces '("CNCLD" . +org-todo-cancel))
  (add-to-list 'org-todo-keyword-faces '("KILLED" . +org-todo-cancel))
  (setq org-tag-alist '(("@work") ("@home") ("@phone") ("@computer") ("@online") ("@errand")))
  (setq org-gtd-inbox-file "~/org/gtd/0-inbox.org")
  (add-to-list 'org-capture-templates
               '("i" "Inbox" entry
                 (file org-gtd-inbox-file)
                 "* %?\n%i\n%a"))
  (add-to-list 'org-capture-templates
               '("p" "Protocol" entry (file org-gtd-inbox-file)
                 "* %^{Title}\nSource: %u, %c\n #+BEGIN_QUOTE\n%i\n#+END_QUOTE\n\n\n%?") t)
  (add-to-list 'org-capture-templates
               '("L" "Protocol Link" entry (file org-gtd-inbox-file)
                 "* %? [[%:link][%:description]] \nCaptured On: %U") t)

  (delete '("p" "Templates for projects")
          org-capture-templates)
  (delete '("pn" "Project-local notes" entry
            (file+headline +org-capture-project-notes-file "Inbox")
            "* %U %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete '("n" "Personal notes" entry
            (file+headline +org-capture-notes-file "Inbox")
            "* %u %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete '("pt" "Project-local todo" entry
            (file+headline +org-capture-project-todo-file "Inbox")
            "* TODO %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete '("oc" "Project changelog" entry #'+org-capture-central-project-changelog-file "* %U %?\n %i\n %a" :heading "Changelog" :prepend t)
          org-capture-templates)
  (delete '("on" "Project notes" entry #'+org-capture-central-project-notes-file "* %U %?\n %i\n %a" :heading "Notes" :prepend t)
          org-capture-templates)
  (delete '("o" "Centralized templates for projects")
          org-capture-templates)
  (delete '("t" "Personal todo" entry
            (file+headline +org-capture-todo-file "Inbox")
            "* [ ] %?\n%i\n%a" :prepend t)
          org-capture-templates)
  (delete  '("pc" "Project-local changelog" entry
             (file+headline +org-capture-project-changelog-file "Unreleased")
             "* %U %?\n%i\n%a" :prepend t)
           org-capture-templates)
  (delete '("ot" "Project todo" entry #'+org-capture-central-project-todo-file "* TODO %?\n %i\n %a" :heading "Tasks" :prepend nil)
          org-capture-templates))
(setq org-agenda-custom-commands
      '(("n" "Agenda and all NEXT tasks"
         ((agenda "" nil)
          (todo "NEXT" nil))
         nil)))
(add-to-list 'org-agenda-custom-commands
             '("d" "Weekly review"
               agenda ""
               ((org-agenda-start-day "-14d")
                (org-agenda-span 14)
                (org-agenda-start-on-weekday 1)
                (org-agenda-start-with-log-mode '(closed))
                (org-agenda-skip-function))))
(defvar org-gtd-archive-file "~/org/gtd/_gtd_archive_2024")
(setq org-archive-location (concat org-gtd-archive-file "::datetree/"))
(setq org-log-done 'time)
(use-package! denote
  :defer t
  :config
 (message "FEATURE LOADED")
  (setq denote-directory org-directory))
(map! :leader "n d" nil)
(map! :leader
      :prefix ("n d" . "denote"))
(map! :leader
      "n d d" #'denote)

(add-hook 'after-save-hook
          'executable-make-buffer-file-executable-if-script-p)
(setq bookmark-default-file "~/.config/doom/bookmarks" )

(setq projectile-project-search-path '(( "~/Repos/" . 1 )))
(after! orderless
  (add-to-list 'completion-styles 'flex t))
(after! corfu
  (setq corfu-preselect 'directory))

(require 'notmuch-mua)
(global-set-key [remap compose-mail] #'+notmuch/compose)
(after! notmuch
  (set-popup-rule! "^\\*notmuch-hello" :ignore t)
  (setq notmuch-multipart/alternative-discouraged '("text/plain" "multipart/related"))
  (setq +notmuch-mail-folder "~/.mail/")
  (setq sendmail-program "gmi")
  (setq message-sendmail-extra-arguments '("send" "--quiet" "-t" "-C" "~/.mail"))
  (setq send-mail-function 'sendmail-send-it)
  (setq message-send-mail-function 'message-send-mail-with-sendmail)
  (require 'notmuch-address)
  (setq notmuch-address-command "~/.scripts/gook")

  (setq notmuch-saved-searches
        '((:name "inbox" :query "tag:inbox" :key "i")
          (:name "unread" :query "tag:unread" :key "u")
          (:name "flagged" :query "tag:flagged" :key "f")
          (:name "sent" :query "tag:sent" :key "t")
          (:name "drafts" :query "tag:draft" :key "d")
          (:name "all" :query "not tag:trash and not tag:deleted and not tag:spam and not tag:sent" :key "a" :search-type tree)))
  (map!
   :map (notmuch-search-mode-map notmuch-tree-mode-map notmuch-show-mode-map)
   :n "K" #'notmuch-tag-jump
   :nv [remap evil-undo] #'notmuch-tag-undo)

  (defun +notmuch/search-trash ()
    "Flag the message."
    (interactive)
    (notmuch-search-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "trash" "search" #'notmuch-tree-next-message))
  (defun +notmuch/show-trash ()
    "Flag the message."
    (interactive)
    (notmuch-show-remove-tag'("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "trash" "show" #'notmuch-show-next-thread-show))
  (defun +notmuch/tree-trash ()
    "Flag the message."
    (interactive)
    (notmuch-tree-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "trash" "tree"))
  (global-set-key [remap evil-collection-notmuch-tree-toggle-delete] #'+notmuch/tree-trash)
  (global-set-key [remap evil-collection-notmuch-show-toggle-delete] #'+notmuch/show-trash)
  (global-set-key [remap evil-collection-notmuch-search-toggle-delete] #'+notmuch/search-trash)
  (defun +notmuch/search-flagged ()
    "Flag the message."
    (interactive)
    (notmuch-search-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "flagged" "search" #'notmuch-tree-next-message))
  (defun +notmuch/show-flagged ()
    "Flag the message."
    (interactive)
    (notmuch-show-remove-tag'("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "flagged" "show" #'notmuch-show-next-thread-show))
  (defun +notmuch/tree-flagged ()
    "Flag the message."
    (interactive)
    (notmuch-tree-remove-tag '("-inbox" "-unread"))
    (evil-collection-notmuch-toggle-tag "flagged" "tree"))
  (global-set-key [remap evil-collection-notmuch-tree-toggle-flagged] #'+notmuch/tree-flagged)
  (global-set-key [remap evil-collection-notmuch-show-toggle-flagged] #'+notmuch/show-flagged)
  (global-set-key [remap evil-collection-notmuch-search-toggle-flagged] #'+notmuch/search-flagged))

(use-package! jinx
  :general ([remap ispell-word] #'jinx-correct)
  :init
  (after! evil
    (global-set-key [remap evil-next-flyspell-error] #'jinx-next)
    (global-set-key [remap evil-prev-flyspell-error] #'jinx-previous))
  (setq jinx-languages "fr_FR en_US"))

(add-hook! (prog-mode text-mode) #'jinx-mode )

(after! evil
  (defun +evil-paste-above ()
    "Paste the line above."
    (interactive)
    (evil-insert-newline-above)
    (evil-paste-after 1 evil-this-register))

  (defun +evil-paste-below ()
    "Paste the line below."
    (interactive)
    (evil-insert-newline-below)
    (evil-paste-after 1 evil-this-register))
  (map!
   :map evil-normal-state-map
   "[ p" #'+evil-paste-above
   "] p" #'+evil-paste-below ))

(setq org-contacts-files `(,(expand-file-name (concat org-directory "contacts.org"))
                           ,(expand-file-name (concat org-directory "contacts-maman.org"))))
(setq org-contacts-matcher
      "N<>\"\"|EMAIL<>\"\"|ALIAS<>\"\"|PHONE<>\"\"|ADDRESS<>\"\"|BIRTHDAY<>\"\"")
(add-load-path! "~/.config/doom/lisp/")
(require 'gtd)
(use-package! org-contacts
  :commands org-contacts-agenda)
(map!
 :leader
 :desc "Org contacts" "o c" #'org-contacts-agenda)

(map! :map search-map
      "g" #'fd-grep-dired
      "n" #'fd-name-dired)
(map! :map dired-mode-map
      :n "K" #'dired-do-kill-lines)

;; Bind your key
;; Optionally re-bind documentation to different key:
(map! :nv "gK"  #'+lookup/documentation)
(after! smartparens
  (require 'hydra)
  (defhydra hydra-smartparens (:hint nil)
    "
 Moving^^^^                       Slurp & Barf^^   Wrapping^^            Sexp juggling^^^^               Destructive
------------------------------------------------------------------------------------------------------------------------
 [_a_] beginning  [_n_] down      [_h_] bw slurp   [_R_]   rewrap        [_S_] split   [_t_] transpose   [_c_] change inner  [_w_] copy
 [_e_] end        [_N_] bw down   [_H_] bw barf    [_u_]   unwrap        [_s_] splice  [_A_] absorb      [_C_] change outer
 [_f_] forward    [_p_] up        [_l_] slurp      [_U_]   bw unwrap     [_r_] raise   [_E_] emit        [_k_] kill          [_g_] quit
 [_b_] backward   [_P_] bw up     [_L_] barf       [_(__{__[_] wrap (){}[]   [_j_] join    [_o_] convolute   [_K_] bw kill       [_q_] quit"
    ;; Moving
    ("a" sp-beginning-of-sexp)
    ("e" sp-end-of-sexp)
    ("f" sp-forward-sexp)
    ("b" sp-backward-sexp)
    ("n" sp-down-sexp)
    ("N" sp-backward-down-sexp)
    ("p" sp-up-sexp)
    ("P" sp-backward-up-sexp)

    ;; Slurping & barfing
    ("h" sp-backward-slurp-sexp)
    ("H" sp-backward-barf-sexp)
    ("l" sp-forward-slurp-sexp)
    ("L" sp-forward-barf-sexp)

    ;; Wrapping
    ("R" sp-rewrap-sexp)
    ("u" sp-unwrap-sexp)
    ("U" sp-backward-unwrap-sexp)
    ("(" sp-wrap-round)
    ("{" sp-wrap-curly)
    ("[" sp-wrap-square)

    ;; Sexp juggling
    ("S" sp-split-sexp)
    ("s" sp-splice-sexp)
    ("r" sp-raise-sexp)
    ("j" sp-join-sexp)
    ("t" sp-transpose-sexp)
    ("A" sp-absorb-sexp)
    ("E" sp-emit-sexp)
    ("o" sp-convolute-sexp)

    ;; Destructive editing
    ("c" sp-change-inner :exit t)
    ("C" sp-change-enclosing :exit t)
    ("k" sp-kill-sexp)
    ("K" sp-backward-kill-sexp)
    ("w" sp-copy-sexp)

    ("q" nil)
    ("g" nil)))
(map! :map prog-mode-map
      :nv "z p" #'hydra-smartparens/body)
;; TODO:  add keybinds
(after! magit
  (setq magit-diff-refine-hunk 'all))
(setq magit-todos-rg-extra-args '("--hidden"))

(map!
        :map corfu-map
        :i "C-g" #'corfu-quit)

(defun doom-project-commander (dir)
  "Open projectile-commander for projects in DIR.

If DIR is not a project, it will be indexed (but not cached)."
  (unless (file-directory-p dir)
    (error "Directory %S does not exist" dir))
  (unless (file-readable-p dir)
    (error "Directory %S isn't readable" dir))
  (projectile-commander))
(setq! +workspaces-switch-project-function 'doom-project-commander)
